File: mobile/modules/bluetooth-sdk/android/src/main/java/com/mentra/bluetoothsdk/ObservableStore.kt
package com.mentra.bluetoothsdk

import java.util.UUID
import org.json.JSONObject

/** Observable state management with immediate event emission */
class ObservableStore {
    companion object {
        const val BLUETOOTH_CATEGORY = "bluetooth"
        private const val LEGACY_CORE_CATEGORY = "core"

        fun normalizeCategory(category: String): String =
                if (category == LEGACY_CORE_CATEGORY) BLUETOOTH_CATEGORY else category
    }

    private val values = mutableMapOf<String, Any>()
    private val emitListeners = linkedMapOf<String, (String, Map<String, Any>) -> Unit>()

    @Synchronized
    fun configure(onEmit: (String, Map<String, Any>) -> Unit) {
        emitListeners["default"] = onEmit
    }

    @Synchronized
    fun addListener(onEmit: (String, Map<String, Any>) -> Unit): String {
        val id = UUID.randomUUID().toString()
        emitListeners[id] = onEmit
        return id
    }

    @Synchronized
    fun removeListener(id: String) {
        emitListeners.remove(id)
    }

    fun set(category: String, key: String, value: Any) {
        val normalizedCategory: String
        val listeners: List<(String, Map<String, Any>) -> Unit>

        synchronized(this) {
            normalizedCategory = normalizeCategory(category)
            val fullKey = "$normalizedCategory.$key"
            val oldValue = values[fullKey]

            // Skip if unchanged
            if (oldValue != null && toJson(oldValue) == toJson(value)) return

            values[fullKey] = value
            listeners = emitListeners.values.toList()
        }

        // Emit immediately, outside the store lock so callbacks can safely re-enter the store.
        listeners.forEach { it(normalizedCategory, mapOf(key to value)) }
    }

    fun remove(category: String, key: String) {
        val normalizedCategory: String
        val updatedSnapshot: Map<String, Any>
        val listeners: List<(String, Map<String, Any>) -> Unit>

        synchronized(this) {
            normalizedCategory = normalizeCategory(category)
            val fullKey = "$normalizedCategory.$key"
            if (!values.containsKey(fullKey)) return // already absent, no notification needed
            values.remove(fullKey)
            val prefix = "$normalizedCategory."
            updatedSnapshot = values.filterKeys { it.startsWith(prefix) }
                    .mapKeys { it.key.removePrefix(prefix) }
            listeners = emitListeners.values.toList()
        }

        // Emit the full updated category snapshot so UI listeners can clear the removed key.
        listeners.forEach { it(normalizedCategory, updatedSnapshot) }
    }

    @Synchronized
    fun get(category: String, key: String): Any? = values["${normalizeCategory(category)}.$key"]

    @Synchronized
    fun getCategory(category: String): Map<String, Any> {
        val prefix = "${normalizeCategory(category)}."
        return values.filterKeys { it.startsWith(prefix) }.mapKeys { it.key.removePrefix(prefix) }
    }

    private fun toJson(value: Any): String {
        return JSONObject(mapOf("v" to value)).toString()
    }
}

File: mobile/modules/bluetooth-sdk/ios/Source/ObservableStore.swift
//
//  ObservableStore.swift
//  BluetoothSdk
//
//  Observable state management with immediate event emission
//

import Foundation

@MainActor
class ObservableStore {
    private var values: [String: Any] = [:]
    private var onEmit: ((String, [String: Any]) -> Void)?
    private var listeners: [String: (String, [String: Any]) -> Void] = [:]

    nonisolated static let bluetoothCategory = "bluetooth"
    private nonisolated static let legacyCoreCategory = "core"

    nonisolated static func normalizeCategory(_ category: String) -> String {
        category == legacyCoreCategory ? bluetoothCategory : category
    }

    func configure(onEmit: @escaping (String, [String: Any]) -> Void) {
        self.onEmit = onEmit
    }

    func addListener(_ listener: @escaping (String, [String: Any]) -> Void) -> String {
        let id = UUID().uuidString
        listeners[id] = listener
        return id
    }

    func removeListener(_ id: String) {
        listeners.removeValue(forKey: id)
    }

    func set(_ category: String, _ key: String, _ value: Any) {
        let normalizedCategory = Self.normalizeCategory(category)
        let fullKey = "\(normalizedCategory).\(key)"
        let oldValue = values[fullKey]

        // Skip if unchanged
        if let old = oldValue, areEqual(old, value) {
            return
        }

        values[fullKey] = value

        // Emit immediately
        let changes = [key: value]
        onEmit?(normalizedCategory, changes)
        for listener in Array(listeners.values) {
            listener(normalizedCategory, changes)
        }
    }

    func remove(_ category: String, _ key: String) {
        let normalizedCategory = Self.normalizeCategory(category)
        let fullKey = "\(normalizedCategory).\(key)"
        guard values[fullKey] != nil else { return }
        values.removeValue(forKey: fullKey)
        // Emit updated category snapshot so UI listeners clear the removed key
        let snapshot = getCategory(normalizedCategory)
        onEmit?(normalizedCategory, snapshot)
        for listener in Array(listeners.values) { listener(normalizedCategory, snapshot) }
    }

    func get(_ category: String, _ key: String) -> Any? {
        values["\(Self.normalizeCategory(category)).\(key)"]
    }

    func wouldSkipSet(_ category: String, _ key: String, _ value: Any) -> Bool {
        let fullKey = "\(Self.normalizeCategory(category)).\(key)"
        guard let oldValue = values[fullKey] else { return false }
        return areEqual(oldValue, value)
    }

    func getCategory(_ category: String) -> [String: Any] {
        var result: [String: Any] = [:]
        let prefix = "\(Self.normalizeCategory(category))."
        for (key, value) in values where key.hasPrefix(prefix) {
            let shortKey = String(key.dropFirst(prefix.count))
            result[shortKey] = value
        }
        return result
    }

    /// Helper to compare values
    private func areEqual(_ lhs: Any, _ rhs: Any) -> Bool {
        if let l = lhs as? String, let r = rhs as? String { return l == r }
        if let l = lhs as? Int, let r = rhs as? Int { return l == r }
        if let l = lhs as? Bool, let r = rhs as? Bool { return l == r }
        if let l = lhs as? Double, let r = rhs as? Double { return l == r }
        if let l = lhs as? [String], let r = rhs as? [String] { return l == r }
        if let l = lhs as? [[String: Any]], let r = rhs as? [[String: Any]] {
            return toJson(l) == toJson(r)
        }
        return false
    }

    private func toJson(_ value: Any) -> String? {
        guard let data = try? JSONSerialization.data(withJSONObject: value) else { return nil }
        return String(data: data, encoding: .utf8)
    }
}
